File: src/vatByJurisdiction.js
"use strict";

const SUPPORTED_DIRECTIONS = new Set(["sale", "purchase", "output", "input"]);

/**
 * Generate VAT payable totals grouped by jurisdiction.
 *
 * @param {Array<{
 *   jurisdiction: string,
 *   direction: "sale" | "purchase" | "output" | "input",
 *   vatAmount?: number,
 *   amount?: number,
 *   vatRate?: number
 * }>} transactions
 * @returns {Array<{
 *   jurisdiction: string,
 *   outputVat: number,
 *   inputVat: number,
 *   vatPayable: number
 * }>}
 */
function calculateVatPayableByJurisdiction(transactions) {
  if (!Array.isArray(transactions)) {
    throw new TypeError("transactions must be an array");
  }

  const aggregate = new Map();

  for (const transaction of transactions) {
    if (!transaction || typeof transaction !== "object") {
      throw new TypeError("each transaction must be an object");
    }

    validateJurisdiction(transaction.jurisdiction);
    const jurisdiction = transaction.jurisdiction.trim();
    const direction = normalizeDirection(transaction.direction);
    const vatAmount = getVatAmount(transaction);

    const bucket = aggregate.get(jurisdiction) ?? { outputVat: 0, inputVat: 0 };
    if (direction === "sale") {
      bucket.outputVat += vatAmount;
    } else {
      bucket.inputVat += vatAmount;
    }

    aggregate.set(jurisdiction, bucket);
  }

  return Array.from(aggregate.entries())
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([jurisdiction, totals]) => ({
      jurisdiction,
      outputVat: roundCurrency(totals.outputVat),
      inputVat: roundCurrency(totals.inputVat),
      vatPayable: roundCurrency(totals.outputVat - totals.inputVat),
    }));
}

function validateJurisdiction(jurisdiction) {
  if (typeof jurisdiction !== "string" || jurisdiction.trim().length === 0) {
    throw new Error("jurisdiction is required and must be a non-empty string");
  }
}

function normalizeDirection(direction) {
  if (typeof direction !== "string") {
    throw new TypeError("direction must be a string");
  }

  const normalized = direction.toLowerCase();
  if (!SUPPORTED_DIRECTIONS.has(normalized)) {
    throw new Error(
      `Unsupported direction "${direction}". Use one of: ${Array.from(SUPPORTED_DIRECTIONS).join(", ")}`
    );
  }

  return normalized === "output" ? "sale" : normalized === "input" ? "purchase" : normalized;
}

function getVatAmount(transaction) {
  if (transaction.vatAmount != null) {
    return toFiniteNumber(transaction.vatAmount, "vatAmount");
  }

  const amount = toFiniteNumber(transaction.amount, "amount");
  if (amount < 0) {
    throw new RangeError(`amount cannot be negative. Received: ${amount}`);
  }

  return amount * toNormalizedRate(transaction.vatRate);
}

function toNormalizedRate(rawRate) {
  if (rawRate == null) {
    return 0;
  }

  const rate = toFiniteNumber(rawRate, "vatRate");
  if (rate < 0) {
    throw new RangeError(`vatRate cannot be negative. Received: ${rate}`);
  }

  // Accept both decimal rates (0.2) and percentage rates (20).
  return rate > 1 ? rate / 100 : rate;
}

function toFiniteNumber(value, fieldName) {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new TypeError(`${fieldName} must be a finite number`);
  }

  return value;
}

function roundCurrency(value) {
  return Math.round((value + Number.EPSILON) * 100) / 100;
}

module.exports = {
  calculateVatPayableByJurisdiction,
  toNormalizedRate,
};

File: vat_by_jurisdiction.py
"""Utilities for computing VAT payable by jurisdiction.

The main entry point is :func:`generate_vat_payable_by_jurisdiction`.
"""

from __future__ import annotations

import argparse
import json
from collections import defaultdict
from decimal import Decimal, ROUND_HALF_UP
from pathlib import Path
from typing import Any, Iterable

TWOPLACES = Decimal("0.01")
OUTPUT_TYPES = {"sale", "output"}
INPUT_TYPES = {"purchase", "input"}


def _to_decimal(value: Any, field_name: str) -> Decimal:
    """Convert a user provided number into a Decimal."""
    try:
        return Decimal(str(value))
    except Exception as exc:  # pragma: no cover - defensive conversion path
        raise ValueError(f"Invalid numeric value for '{field_name}': {value!r}") from exc


def _quantize_money(value: Decimal) -> Decimal:
    """Round to two decimal places using standard commercial rounding."""
    return value.quantize(TWOPLACES, rounding=ROUND_HALF_UP)


def _extract_vat_amount(transaction: dict[str, Any]) -> Decimal:
    """Read VAT amount directly, or derive from amount and VAT rate."""
    if "vat_amount" in transaction:
        vat = _to_decimal(transaction["vat_amount"], "vat_amount")
    else:
        if "amount" not in transaction or "vat_rate" not in transaction:
            raise ValueError(
                "Each transaction must include 'vat_amount' or both "
                "'amount' and 'vat_rate'."
            )
        amount = _to_decimal(transaction["amount"], "amount")
        vat_rate = _to_decimal(transaction["vat_rate"], "vat_rate")
        vat = amount * vat_rate
    return _quantize_money(vat)


def generate_vat_payable_by_jurisdiction(
    transactions: Iterable[dict[str, Any]],
) -> dict[str, dict[str, float]]:
    """Generate VAT payable for each jurisdiction.

    Args:
        transactions: Iterable of transaction dictionaries. Each transaction
            must include:
              - ``jurisdiction``: jurisdiction name/code (str)
              - ``type``: one of ``sale``/``output`` or ``purchase``/``input``
              - Either ``vat_amount`` OR both ``amount`` and ``vat_rate``

    Returns:
        A dictionary keyed by jurisdiction with values:
            - ``output_vat``: VAT collected on sales
            - ``input_vat``: VAT paid on purchases
            - ``vat_payable``: VAT due for payment (never negative)
            - ``vat_credit``: recoverable VAT credit (never negative)
    """
    output_totals: dict[str, Decimal] = defaultdict(lambda: Decimal("0"))
    input_totals: dict[str, Decimal] = defaultdict(lambda: Decimal("0"))

    for index, transaction in enumerate(transactions):
        if not isinstance(transaction, dict):
            raise ValueError(f"Transaction at index {index} must be a dictionary.")

        jurisdiction = transaction.get("jurisdiction")
        if not jurisdiction or not isinstance(jurisdiction, str):
            raise ValueError(
                f"Transaction at index {index} is missing a valid 'jurisdiction'."
            )

        tx_type_raw = transaction.get("type")
        if not isinstance(tx_type_raw, str):
            raise ValueError(
                f"Transaction at index {index} is missing a valid 'type'."
            )
        tx_type = tx_type_raw.lower()

        vat_amount = _extract_vat_amount(transaction)
        if vat_amount < 0:
            raise ValueError(f"Transaction at index {index} has a negative VAT amount.")

        if tx_type in OUTPUT_TYPES:
            output_totals[jurisdiction] += vat_amount
        elif tx_type in INPUT_TYPES:
            input_totals[jurisdiction] += vat_amount
        else:
            raise ValueError(
                f"Transaction at index {index} has unsupported type '{tx_type_raw}'. "
                "Use 'sale'/'output' or 'purchase'/'input'."
            )

    jurisdictions = sorted(set(output_totals) | set(input_totals))
    result: dict[str, dict[str, float]] = {}
    for jurisdiction in jurisdictions:
        output_vat = _quantize_money(output_totals[jurisdiction])
        input_vat = _quantize_money(input_totals[jurisdiction])
        net = _quantize_money(output_vat - input_vat)

        vat_payable = net if net > 0 else Decimal("0")
        vat_credit = -net if net < 0 else Decimal("0")

        result[jurisdiction] = {
            "output_vat": float(output_vat),
            "input_vat": float(input_vat),
            "vat_payable": float(vat_payable),
            "vat_credit": float(vat_credit),
        }

    return result


def _parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Generate VAT payable by jurisdiction from JSON transactions."
    )
    parser.add_argument(
        "input_file",
        type=Path,
        help="Path to JSON file containing a list of transactions.",
    )
    return parser.parse_args()


def main() -> None:
    args = _parse_args()
    with args.input_file.open("r", encoding="utf-8") as infile:
        transactions = json.load(infile)
    output = generate_vat_payable_by_jurisdiction(transactions)
    print(json.dumps(output, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
